<!DOCTYPE html>
<head>
    <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta name="description" content="">
<meta name="keywords" content=" ">
<title>Introduction to R | Jekyll theme for documentation</title>
<link rel="stylesheet" href="css/syntax.css">


<link rel="stylesheet" type="text/css" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css">
<!--<link rel="stylesheet" type="text/css" href="css/bootstrap.min.css">-->
<link rel="stylesheet" href="css/modern-business.css">
<link rel="stylesheet" href="css/lavish-bootstrap.css">
<link rel="stylesheet" href="css/customstyles.css">
<link rel="stylesheet" href="css/theme-blue.css">

<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery-cookie/1.4.1/jquery.cookie.min.js"></script>
<script src="js/jquery.navgoco.min.js"></script>


<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.4/js/bootstrap.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/anchor-js/2.0.0/anchor.min.js"></script>
<script src="js/toc.js"></script>
<script src="js/customscripts.js"></script>

<link rel="shortcut icon" href="images/favicon.ico">

<!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
<!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
<script src="https://oss.maxcdn.com/libs/respond.js/1.4.2/respond.min.js"></script>
<![endif]-->

<link rel="alternate" type="application/rss+xml" title="" href="feed.xml">

    <script>
        $(document).ready(function() {
            // Initialize navgoco with default options
            $("#mysidebar").navgoco({
                caretHtml: '',
                accordion: true,
                openClass: 'active', // open
                save: false, // leave false or nav highlighting doesn't work right
                cookie: {
                    name: 'navgoco',
                    expires: false,
                    path: '/'
                },
                slide: {
                    duration: 400,
                    easing: 'swing'
                }
            });

            $("#collapseAll").click(function(e) {
                e.preventDefault();
                $("#mysidebar").navgoco('toggle', false);
            });

            $("#expandAll").click(function(e) {
                e.preventDefault();
                $("#mysidebar").navgoco('toggle', true);
            });

        });

    </script>
    <script>
        $(function () {
            $('[data-toggle="tooltip"]').tooltip()
        })
    </script>
    

</head>
<body>
<!-- Navigation -->
<nav class="navbar navbar-inverse navbar-fixed-top">
    <div class="container topnavlinks">
        <div class="navbar-header">
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            <a class="fa fa-home fa-lg navbar-brand" href="index.html">&nbsp;<span class="projectTitle"> Jekyll Documentation Theme</span></a>
        </div>
        <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
            <ul class="nav navbar-nav navbar-right">
                <!-- entries without drop-downs appear here -->
                
                
                
                <li><a href="news">News</a></li>
                
                
                
                <!-- entries with drop-downs appear here -->
                <!-- conditional logic to control which topnav appears for the audience defined in the configuration file.-->
                
                
                <li class="dropdown">
                    <a href="#" class="dropdown-toggle" data-toggle="dropdown">Jekyll Resources<b class="caret"></b></a>
                    <ul class="dropdown-menu">
                        
                        
                        <li><a href="https://github.com/tomjohnson1492/documentation-theme-jekyll" target="_blank">Github Repo</a></li>
                        
                        
                        
                        <li><a href="https://talk.jekyllrb.com" target="_blank">Jekyll Talk</a></li>
                        
                        
                        
                        <li><a href="http://jekyllrb.com/docs/home/" target="_blank">Jekyll documentation</a></li>
                        
                        
                        
                        <li><a href="http://stackoverflow.com/questions/tagged/jekyll" target="_blank">Jekyll on Stack Overflow</a></li>
                        
                        
                        
                        <li><a href="http://idratherbewriting.com/category-jekyll/" target="_blank">Jekyll on my blog</a></li>
                        
                        
                    </ul>
                </li>
                
                <li class="dropdown">
                    <a href="#" class="dropdown-toggle" data-toggle="dropdown">Products<b class="caret"></b></a>
                    <ul class="dropdown-menu">
                        
                        
                        <li><a href="mydoc_introduction.html">Jekyll Documentation Theme</a></li>
                        
                        
                        
                        <li><a href="p1_landing_page.html">Product 1</a></li>
                        
                        
                        
                        <li><a href="p2_landing_page.html">Product 2</a></li>
                        
                        
                    </ul>
                </li>
                
                
                
			<!-- Send feedback function -->
<script>
function SendLinkByMail(href) {
var subject= "Jekyll theme for documentation feedback";
var body = "I have some feedback about the Introduction to R page: ";
body += window.location.href;
body += "";
var uri = "mailto:tomjohnson1492@gmail.com?subject=";
uri += encodeURIComponent(subject);
uri += "&body=";
uri += encodeURIComponent(body);
window.location.href = uri;
}
</script>

<li><a href="javascript:(function()%7BSendLinkByMail()%3B%7D)()%3B" target="_blank"><i class="fa fa-envelope-o"></i> Feedback</a></li>

		
                <!--comment out this block if you want to hide search-->
                <li>
                    <!--start search-->
                    <div id="search-demo-container">
                        <input type="text" id="search-input" placeholder="search...">
                        <ul id="results-container"></ul>
                    </div>
                    <script src="js/jekyll-search.js" type="text/javascript"></script>
                    <script type="text/javascript">
                            SimpleJekyllSearch.init({
                                searchInput: document.getElementById('search-input'),
                                resultsContainer: document.getElementById('results-container'),
                                dataSource: '../search.json',
                                searchResultTemplate: '<li><a href="{url}" title="Introduction to R">{title}</a></li>',
                    noResultsText: 'No results found.',
                            limit: 10,
                            fuzzy: true,
                    })
                    </script>
                    <!--end search-->
                </li>
            </ul>
        </div>
        </div>
        <!-- /.container -->
</nav>

<!-- Page Content -->
<div class="container">
    <div class="col-lg-12">&nbsp;</div>
    <!-- Content Row -->
    <div class="row">
        <!-- Sidebar Column -->
        <div class="col-md-3">

          












<ul id="mysidebar" class="nav">
    <li class="sidebarTitle"> </li>
    
    
    
    <li>
        <a href="#">Products</a>
        <ul>
            
            
            
            <li><a href="news.html">News</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_introduction.html">Theme instructions</a></li>
            
            
            
            
            
            
            <li><a href="p1_landing_page.html">Product 1</a></li>
            
            
            
            
            
            
            <li><a href="p2_landing_page.html">Product 2</a></li>
            
            
            
            
        </ul>
        
        
        
        <!-- if you aren't using the accordion, uncomment this block:
           <p class="external">
               <a href="#" id="collapseAll">Collapse All</a> | <a href="#" id="expandAll">Expand All</a>
           </p>
           -->
    </li>
</ul>
</div>

<!-- this highlights the active parent class in the navgoco sidebar. this is critical so that the parent expands when you're viewing a page. This must appear below the sidebar code above. Otherwise, if placed inside customscripts.js, the script runs before the sidebar code runs and the class never gets inserted.-->
<script>$("li.active").parents('li').toggleClass("active");</script>

    <!-- Content Column -->
    <div class="col-md-9">
        <div class="post-header">
   <h1 class="post-title-main">Introduction to R</h1>
</div>



<div class="post-content">

   

    
    
<!-- this handles the automatic toc. use ## for subheads to auto-generate the on-page minitoc. if you use html tags, you must supply an ID for the heading element in order for it to appear in the minitoc. -->
<script>
$( document ).ready(function() {
  // Handler for .ready() called.

$('#toc').toc({ minimumHeaders: 0, listType: 'ul', showSpeed: 0, headers: 'h2,h3,h4' });

/* this offset helps account for the space taken up by the floating toolbar. */
$('#toc').on('click', 'a', function() {
  var target = $(this.getAttribute('href'))
    , scroll_target = target.offset().top

  $(window).scrollTop(scroll_target - 10);
  return false
})
  
});
</script>

<div id="toc"></div>

    

    

    

    <a target="_blank" href="https://github.com/tomjohnson1492/documentation-theme-jekyll/blob/gh-pages/_pages//_pages//vignettes07_RbasicsRbasics.knit.html.md" class="btn btn-default githubEditButton" role="button"><i class="fa fa-github fa-lg"></i> Edit me</a>
    

    

  <!--
%% \VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Overview Vignette}
%% \VignetteDepends{methods}
%% \VignetteKeywords{compute cluster, pipeline, reports}
%% \VignettePackage{longevityTools}
-->

<!---
- Compile from command-line
echo "rmarkdown::render('Rbasics.Rmd', clean=F)" | R -slave; R CMD Stangle Rbasics.Rmd; Rscript ../md2jekyll.R Rbasics.knit.md 8

- Commit to github
git commit -am "some edits"; git push -u origin master

- To customize font size and other style features, add this line to output section in preamble:  
    css: style.css
-->

<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script>

<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script>

<h1 id="overview">Overview</h1>

<h2 id="what-is-r">What is R?</h2>

<p><a href="http://cran.at.r-project.org">R</a> is a powerful statistical environment and
programming language for the analysis and visualization of data.  The
associated <a href="http://bioconductor.org/">Bioconductor</a> and CRAN package
repositories provide many additional R packages for statistical data analysis
for a wide array of research areas. The R software is free and runs on all
common operating systems.</p>

<h2 id="why-using-r">Why Using R?</h2>
<ul>
  <li>Complete statistical environment and programming language</li>
  <li>Efficient functions and data structures for data analysis</li>
  <li>Powerful graphics</li>
  <li>Access to fast growing number of analysis packages</li>
  <li>Most widely used language in bioinformatics</li>
  <li>Is standard for data mining and biostatistical analysis</li>
  <li>Technical advantages: free, open-source, available for all OSs</li>
</ul>

<h2 id="books-and-documentation">Books and Documentation</h2>
<ul>
  <li>simpleR - Using R for Introductory Statistics (John Verzani, 2004) - <a href="http://cran.r-project.org/doc/contrib/Verzani-SimpleR.pdf">URL</a></li>
  <li>Bioinformatics and Computational Biology Solutions Using R and Bioconductor (Gentleman et al., 2005) - <a href="http://www.bioconductor.org/help/publications/books/bioinformatics-and-computational-biology-solutions/">URL</a></li>
  <li>More on this see “Finding Help” section in UCR Manual - <a href="http://manuals.bioinformatics.ucr.edu/home/R\_BioCondManual\#TOC-Finding-Help">URL</a></li>
</ul>

<h2 id="r-working-environments">R Working Environments</h2>

<h2 id="working-environments-ides-for-r">Working environments (IDEs) for R</h2>
<center><img title="R_Interfaces" src="images/rinterface.png" /></center>
<center> R Projects and Interfaces</center>

<p>Some R working environments with support for syntax highlighting and utilities to send code 
to the R console:</p>

<ul>
  <li><a href="https://www.rstudio.com/products/rstudio/features">RStudio</a>: excellent choice for beginners (<a href="http://www.rstudio.com/wp-content/uploads/2016/01/rstudio-IDE-cheatsheet.pdf">Cheat Sheet</a>)</li>
  <li>Basic R code editors provided by Rguis</li>
  <li><a href="https://wiki.gnome.org/Apps/Gedit">gedit</a>, <a href="http://rgedit.sourceforge.net/">Rgedit</a>, <a href="https://rkward.kde.org/">RKWard</a>, <a href="http://www.walware.de/goto/statet">Eclipse</a>, <a href="http://www.sciviews.org/Tinn-R/">Tinn-R</a>, <a href="https://notepad-plus-plus.org/">Notepad++</a>, <a href="http://sourceforge.net/projects/npptor/">NppToR</a></li>
  <li><a href="http://manuals.bioinformatics.ucr.edu/home/programming-in-r/vim-r">Vim-R-Tmux</a>: R working environment based on vim and tmux</li>
  <li><a href="http://www.xemacs.org/Download/index.html">Emacs</a> (<a href="http://ess.r-project.org/">ESS add-on package</a>)</li>
</ul>

<h3 id="example-rstudio">Example: RStudio</h3>

<p>New integrated development environment (IDE) for <a href="http://www.rstudio.com/ide/download/">R</a>. Highly functional for both beginners and 
advanced.</p>

<center><img title="RStudio" src="images/rstudio.png" /></center>
<center> RStudio IDE</center>

<p>Some userful shortcuts: <code class="highlighter-rouge">Ctrl+Enter</code> (send code), <code class="highlighter-rouge">Ctrl+Shift+C</code> (comment/uncomment), <code class="highlighter-rouge">Ctrl+1/2</code> (switch window focus)</p>

<h3 id="example-vim-r-tmux">Example: Vim-R-Tmux</h3>

<p>Terminal-based Working Environment for R: <a href="http://manuals.bioinformatics.ucr.edu/home/programming-in-r/vim-r">Vim-R-Tmux</a></p>

<center><img title="Vim-R-Tmux" src="images/screenshot.png" /></center>
<center>Vim-R-Tmux IDE for R</center>

<h1 id="r-package-repositories">R Package Repositories</h1>

<ul>
  <li>CRAN (&gt;8,000 packages) general data analysis - <a href="http://cran.at.r-project.org/">URL</a></li>
  <li>Bioconductor (&gt;1,100 packages) bioscience data analysis - <a href="http://www.bioconductor.org/">URL</a></li>
  <li>Omegahat (&gt;90 packages) programming interfaces - <a href="https://github.com/omegahat?tab=repositories">URL</a></li>
</ul>

<h1 id="installation-of-r-and-add-on-packages">Installation of R and Add-on Packages</h1>

<p>(1.) Install R for your operating system from <a href="http://cran.at.r-project.org/">CRAN</a>.</p>

<p>(2.) Install RStudio from <a href="http://www.rstudio.com/ide/download">RStudio</a>.</p>

<p>(3.) Install CRAN Packages from R console like this:</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">install.packages</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s2">"pkg1"</span><span class="p">,</span><span class="w"> </span><span class="s2">"pkg2"</span><span class="p">))</span><span class="w"> 
</span><span class="n">install.packages</span><span class="p">(</span><span class="s2">"pkg.zip"</span><span class="p">,</span><span class="w"> </span><span class="n">repos</span><span class="o">=</span><span class="kc">NULL</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p>(4.) Install Bioconductor packages as follows:</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">source</span><span class="p">(</span><span class="s2">"http://www.bioconductor.org/biocLite.R"</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">BiocInstaller</span><span class="p">)</span><span class="w">
</span><span class="n">BiocVersion</span><span class="p">()</span><span class="w">
</span><span class="n">biocLite</span><span class="p">()</span><span class="w">
</span><span class="n">biocLite</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s2">"pkg1"</span><span class="p">,</span><span class="w"> </span><span class="s2">"pkg2"</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<p>(5.) For more details consult the <a href="http://www.bioconductor.org/install/">Bioc Install page</a>
and <a href="http://www.bioconductor.org/packages/release/bioc/html/BiocInstaller.html">BiocInstaller</a> package.</p>

<h1 id="getting-around">Getting Around</h1>

<h2 id="startup-and-closing-behavior">Startup and Closing Behavior</h2>

<ul>
  <li>
    <p><strong>Starting R</strong>:
  The R GUI versions, including RStudio, under Windows and Mac OS X can be
  opened by double-clicking their icons. Alternatively, one can start it by
  typing <code class="highlighter-rouge">R</code> in a terminal (default under Linux).</p>
  </li>
  <li>
    <p><strong>Startup/Closing Behavior</strong>:
  The R environment is controlled by hidden files in the startup directory:
  <code class="highlighter-rouge">.RData</code>, <code class="highlighter-rouge">.Rhistory</code> and <code class="highlighter-rouge">.Rprofile</code> (optional).</p>
  </li>
  <li>
    <p><strong>Closing R</strong>:</p>
  </li>
</ul>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">q</span><span class="p">()</span><span class="w">  
</span></code></pre>
</div>
<div class="highlighter-rouge"><pre class="highlight"><code>Save workspace image? [y/n/c]:
</code></pre>
</div>

<ul>
  <li><strong>Note</strong>:
  When responding with <code class="highlighter-rouge">y</code>, then the entire R workspace will be written to
  the <code class="highlighter-rouge">.RData</code> file which can become very large. Often it is sufficient to just
  save an analysis protocol in an R source file. This way one can quickly
  regenerate all data sets and objects.</li>
</ul>

<h2 id="navigating-directories">Navigating directories</h2>

<p>Create an object with the assignment operator <code class="highlighter-rouge">&lt;-</code> or <code class="highlighter-rouge">=</code></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">object</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">...</span><span class="w">
</span></code></pre>
</div>

<p>List objects in current R session</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">ls</span><span class="p">()</span><span class="w">
</span></code></pre>
</div>

<p>Return content of current working directory</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">dir</span><span class="p">()</span><span class="w">
</span></code></pre>
</div>

<p>Return path of current working directory</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">getwd</span><span class="p">()</span><span class="w">
</span></code></pre>
</div>

<p>Change current working directory</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">setwd</span><span class="p">(</span><span class="s2">"/home/user"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<h1 id="basic-syntax">Basic Syntax</h1>

<p>General R command syntax</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">object</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">function_name</span><span class="p">(</span><span class="n">arguments</span><span class="p">)</span><span class="w"> 
</span><span class="n">object</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">object</span><span class="p">[</span><span class="n">arguments</span><span class="p">]</span><span class="w"> 
</span></code></pre>
</div>

<p>Finding help</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="o">?</span><span class="n">function_name</span><span class="w">
</span></code></pre>
</div>

<p>Load a library/package</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="s2">"my_library"</span><span class="p">)</span><span class="w"> 
</span></code></pre>
</div>

<p>List functions defined by a library</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">help</span><span class="o">=</span><span class="s2">"my_library"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p>Load library manual (PDF or HTML file)</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">vignette</span><span class="p">(</span><span class="s2">"my_library"</span><span class="p">)</span><span class="w"> 
</span></code></pre>
</div>

<p>Execute an R script from within R</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">source</span><span class="p">(</span><span class="s2">"my_script.R"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p>Execute an R script from command-line (the first of the three options is preferred)</p>

<div class="language-sh highlighter-rouge"><pre class="highlight"><code><span class="gp">$ </span>Rscript my_script.R
<span class="gp">$ </span>R CMD BATCH my_script.R 
<span class="gp">$ </span>R --slave &lt; my_script.R 
</code></pre>
</div>

<h1 id="data-types">Data Types</h1>

<h2 id="numeric-data">Numeric data</h2>

<p>Example: <code class="highlighter-rouge">1, 2, 3, ...</code></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">x</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">3</span><span class="p">)</span><span class="w">
</span><span class="n">x</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 1 2 3
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">is.numeric</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] TRUE
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">as.character</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "1" "2" "3"
</code></pre>
</div>

<h2 id="character-data">Character data</h2>

<p>Example: <code class="highlighter-rouge">"a", "b", "c", ...</code></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">x</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"1"</span><span class="p">,</span><span class="w"> </span><span class="s2">"2"</span><span class="p">,</span><span class="w"> </span><span class="s2">"3"</span><span class="p">)</span><span class="w">
</span><span class="n">x</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "1" "2" "3"
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">is.character</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] TRUE
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">as.numeric</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 1 2 3
</code></pre>
</div>

<h2 id="complex-data">Complex data</h2>

<p>Example: mix of both</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="s2">"b"</span><span class="p">,</span><span class="w"> </span><span class="m">3</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "1" "b" "3"
</code></pre>
</div>

<h2 id="logical-data">Logical data</h2>

<p>Example: <code class="highlighter-rouge">TRUE</code> of <code class="highlighter-rouge">FALSE</code></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">x</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">5</span><span class="w">
</span><span class="n">x</span><span class="w">  
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##  [1]  TRUE  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="o">!</span><span class="n">x</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##  [1] FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">which</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="c1"># Returns index for the 'TRUE' values in logical vector
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 1 2 3 4
</code></pre>
</div>

<h1 id="data-objects">Data objects</h1>

<h2 id="object-types">Object types</h2>

<h2 id="vectors-1d">Vectors (1D)</h2>

<p>Definition: <code class="highlighter-rouge">numeric</code> or <code class="highlighter-rouge">character</code></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">myVec</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">;</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">myVec</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nb">letters</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">]</span><span class="w">  
</span><span class="n">myVec</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## a b c d e 
## 1 2 3 4 5
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">myVec</span><span class="p">[</span><span class="nf">c</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="m">4</span><span class="p">,</span><span class="m">6</span><span class="p">,</span><span class="m">8</span><span class="p">)]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## b d f h 
## 2 4 6 8
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">myVec</span><span class="p">[</span><span class="nf">c</span><span class="p">(</span><span class="s2">"b"</span><span class="p">,</span><span class="w"> </span><span class="s2">"d"</span><span class="p">,</span><span class="w"> </span><span class="s2">"f"</span><span class="p">)]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## b d f 
## 2 4 6
</code></pre>
</div>

<h2 id="factors-1d">Factors (1D)</h2>

<p>Definition: vectors with grouping information</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">factor</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s2">"dog"</span><span class="p">,</span><span class="w"> </span><span class="s2">"cat"</span><span class="p">,</span><span class="w"> </span><span class="s2">"mouse"</span><span class="p">,</span><span class="w"> </span><span class="s2">"dog"</span><span class="p">,</span><span class="w"> </span><span class="s2">"dog"</span><span class="p">,</span><span class="w"> </span><span class="s2">"cat"</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] dog   cat   mouse dog   dog   cat  
## Levels: cat dog mouse
</code></pre>
</div>

<h2 id="matrices-2d">Matrices (2D)</h2>

<p>Definition: two dimensional structures with data of same type</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">myMA</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">matrix</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">30</span><span class="p">,</span><span class="w"> </span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">byrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w"> 
</span><span class="nf">class</span><span class="p">(</span><span class="n">myMA</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "matrix"
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">myMA</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">,]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
## [1,]    1    2    3    4    5    6    7    8    9    10
## [2,]   11   12   13   14   15   16   17   18   19    20
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">myMA</span><span class="p">[</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="p">,</span><span class="w"> </span><span class="n">drop</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
## [1,]    1    2    3    4    5    6    7    8    9    10
</code></pre>
</div>

<h2 id="data-frames-2d">Data Frames (2D)</h2>

<p>Definition: two dimensional objects with data of variable types</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">myDF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">Col1</span><span class="o">=</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">Col2</span><span class="o">=</span><span class="m">10</span><span class="o">:</span><span class="m">1</span><span class="p">)</span><span class="w"> 
</span><span class="n">myDF</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   Col1 Col2
## 1    1   10
## 2    2    9
</code></pre>
</div>

<h2 id="arrays">Arrays</h2>

<p>Definition: data structure with one, two or more dimensions</p>

<h2 id="lists">Lists</h2>

<p>Definition: containers for any object type</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">myL</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">"Fred"</span><span class="p">,</span><span class="w"> </span><span class="n">wife</span><span class="o">=</span><span class="s2">"Mary"</span><span class="p">,</span><span class="w"> </span><span class="n">no.children</span><span class="o">=</span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">child.ages</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">4</span><span class="p">,</span><span class="m">7</span><span class="p">,</span><span class="m">9</span><span class="p">))</span><span class="w"> 
</span><span class="n">myL</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## $name
## [1] "Fred"
## 
## $wife
## [1] "Mary"
## 
## $no.children
## [1] 3
## 
## $child.ages
## [1] 4 7 9
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">myL</span><span class="p">[[</span><span class="m">4</span><span class="p">]][</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">]</span><span class="w"> 
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 4 7
</code></pre>
</div>

<h2 id="functions">Functions</h2>

<p>Definition: piece of code</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">myfct</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">arg1</span><span class="p">,</span><span class="w"> </span><span class="n">arg2</span><span class="p">,</span><span class="w"> </span><span class="n">...</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w"> 
	</span><span class="n">function_body</span><span class="w"> 
</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>

<h2 id="subsetting-of-data-objects">Subsetting of data objects</h2>

<p><strong>(1.) Subsetting by positive or negative index/position numbers</strong></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">myVec</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">26</span><span class="p">;</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">myVec</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nb">LETTERS</span><span class="w"> 
</span><span class="n">myVec</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## A B C D 
## 1 2 3 4
</code></pre>
</div>

<p><strong>(2.) Subsetting by same length logical vectors</strong></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">myLog</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">myVec</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">10</span><span class="w">
</span><span class="n">myVec</span><span class="p">[</span><span class="n">myLog</span><span class="p">]</span><span class="w"> 
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##  K  L  M  N  O  P  Q  R  S  T  U  V  W  X  Y  Z 
## 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26
</code></pre>
</div>

<p><strong>(3.) Subsetting by field names</strong></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">myVec</span><span class="p">[</span><span class="nf">c</span><span class="p">(</span><span class="s2">"B"</span><span class="p">,</span><span class="w"> </span><span class="s2">"K"</span><span class="p">,</span><span class="w"> </span><span class="s2">"M"</span><span class="p">)]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##  B  K  M 
##  2 11 13
</code></pre>
</div>

<p><strong>(4.) Subset with <code class="highlighter-rouge">$</code> sign</strong>: references a single column or list component by its name</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">iris</span><span class="o">$</span><span class="n">Species</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">8</span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] setosa setosa setosa setosa setosa setosa setosa setosa
## Levels: setosa versicolor virginica
</code></pre>
</div>

<h1 id="important-utilities">Important Utilities</h1>

<h2 id="combining-objects">Combining Objects</h2>

<p>The <code class="highlighter-rouge">c</code> function combines vectors and lists</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">3</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 1 2 3
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">x</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">3</span><span class="p">;</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">101</span><span class="o">:</span><span class="m">103</span><span class="w">
</span><span class="nf">c</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1]   1   2   3 101 102 103
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">iris</span><span class="o">$</span><span class="n">Species</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">8</span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] setosa setosa setosa setosa setosa setosa setosa setosa
## Levels: setosa versicolor virginica
</code></pre>
</div>

<p>The <code class="highlighter-rouge">cbind</code> and <code class="highlighter-rouge">rbind</code> functions can be used to append columns and rows, respecively.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">ma</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">cbind</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="p">)</span><span class="w">
</span><span class="n">ma</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##      x   y
## [1,] 1 101
## [2,] 2 102
## [3,] 3 103
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">rbind</span><span class="p">(</span><span class="n">ma</span><span class="p">,</span><span class="w"> </span><span class="n">ma</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##      x   y
## [1,] 1 101
## [2,] 2 102
## [3,] 3 103
## [4,] 1 101
## [5,] 2 102
## [6,] 3 103
</code></pre>
</div>

<h2 id="accessing-dimensions-of-objects">Accessing Dimensions of Objects</h2>

<p>Length and dimension information of objects</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">length</span><span class="p">(</span><span class="n">iris</span><span class="o">$</span><span class="n">Species</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 150
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">dim</span><span class="p">(</span><span class="n">iris</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 150   5
</code></pre>
</div>

<h2 id="accessing-name-slots-of-objects">Accessing Name Slots of Objects</h2>

<p>Accessing row and column names of 2D objects</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">rownames</span><span class="p">(</span><span class="n">iris</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">8</span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "1" "2" "3" "4" "5" "6" "7" "8"
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">colnames</span><span class="p">(</span><span class="n">iris</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"  "Species"
</code></pre>
</div>

<p>Return name field of vectors and lists</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">names</span><span class="p">(</span><span class="n">myVec</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##  [1] "A" "B" "C" "D" "E" "F" "G" "H" "I" "J" "K" "L" "M" "N" "O" "P" "Q" "R" "S" "T" "U" "V" "W" "X"
## [25] "Y" "Z"
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">names</span><span class="p">(</span><span class="n">myL</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "name"        "wife"        "no.children" "child.ages"
</code></pre>
</div>

<h2 id="sorting-objects">Sorting Objects</h2>

<p>The function <code class="highlighter-rouge">sort</code> returns a vector in ascending or descending order</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">sort</span><span class="p">(</span><span class="m">10</span><span class="o">:</span><span class="m">1</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##  [1]  1  2  3  4  5  6  7  8  9 10
</code></pre>
</div>

<p>The function <code class="highlighter-rouge">order</code> returns a sorting index for sorting an object</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">sortindex</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">order</span><span class="p">(</span><span class="n">iris</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="n">decreasing</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span><span class="w">
</span><span class="n">sortindex</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">12</span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##  [1] 14  9 39 43 42  4  7 23 48  3 30 12
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">iris</span><span class="p">[</span><span class="n">sortindex</span><span class="p">,][</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">,]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 14          4.3         3.0          1.1         0.1  setosa
## 9           4.4         2.9          1.4         0.2  setosa
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">sortindex</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">order</span><span class="p">(</span><span class="o">-</span><span class="n">iris</span><span class="p">[,</span><span class="m">1</span><span class="p">])</span><span class="w"> </span><span class="c1"># Same as decreasing=TRUE
</span></code></pre>
</div>
<p>Sorting multiple columns</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">iris</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">iris</span><span class="o">$</span><span class="n">Sepal.Length</span><span class="p">,</span><span class="w"> </span><span class="n">iris</span><span class="o">$</span><span class="n">Sepal.Width</span><span class="p">),][</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">,]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 14          4.3         3.0          1.1         0.1  setosa
## 9           4.4         2.9          1.4         0.2  setosa
</code></pre>
</div>

<h1 id="operators-and-calculations">Operators and Calculations</h1>

<h2 id="comparison-operators">Comparison Operators</h2>

<p>Comparison operators: <code class="highlighter-rouge">==</code>, <code class="highlighter-rouge">!=</code>, <code class="highlighter-rouge">&lt;</code>, <code class="highlighter-rouge">&gt;</code>, <code class="highlighter-rouge">&lt;=</code>, <code class="highlighter-rouge">&gt;=</code></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="m">1</span><span class="o">==</span><span class="m">1</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] TRUE
</code></pre>
</div>
<p>Logical operators: AND: <code class="highlighter-rouge">&amp;</code>, OR: <code class="highlighter-rouge">|</code>, NOT: <code class="highlighter-rouge">!</code></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">x</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">;</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">10</span><span class="o">:</span><span class="m">1</span><span class="w">
</span><span class="n">x</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">5</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##  [1] FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE
</code></pre>
</div>

<h2 id="basic-calculations">Basic Calculations</h2>

<p>To look up math functions, see Function Index <a href="http://cran.at.r-project.org/doc/manuals/R-intro.html#Function-and-variable-index">here</a></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">x</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">y</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##  [1] 11 11 11 11 11 11 11 11 11 11
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">sum</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 55
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">mean</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 5.5
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">apply</span><span class="p">(</span><span class="n">iris</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">6</span><span class="p">,</span><span class="m">1</span><span class="o">:</span><span class="m">3</span><span class="p">],</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">mean</span><span class="p">)</span><span class="w"> 
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##        1        2        3        4        5        6 
## 3.333333 3.100000 3.066667 3.066667 3.333333 3.666667
</code></pre>
</div>

<h1 id="reading-and-writing-external-data">Reading and Writing External Data</h1>
<p>## Import of tabular data</p>

<p>Import of a tab-delimited tabular file</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">myDF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.delim</span><span class="p">(</span><span class="s2">"myData.xls"</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p>Import of Excel file. Note: working with tab- or comma-delimited files is more flexible and preferred.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">gdata</span><span class="p">)</span><span class="w">
</span><span class="n">myDF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.xls</span><span class="s2">"myData.xls"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p>Import of Google Sheets. The following example imports a sample Google Sheet from <a href="https://docs.google.com/spreadsheets/d/1U-32UcwZP1k3saKeaH1mbvEAOfZRdNHNkWK2GI1rpPM/edit#gid=472150521">here</a>.
Detailed instructions for interacting from R with Google Sheets with the required <code class="highlighter-rouge">googlesheets</code> package are <a href="https://github.com/jennybc/googlesheets">here</a>.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="s2">"googlesheets"</span><span class="p">);</span><span class="w"> </span><span class="n">library</span><span class="p">(</span><span class="s2">"dplyr"</span><span class="p">);</span><span class="w"> </span><span class="n">library</span><span class="p">(</span><span class="n">knitr</span><span class="p">)</span><span class="w">
</span><span class="n">gs_auth</span><span class="p">()</span><span class="w"> </span><span class="c1"># Creates authorizaton token (.httr-oauth) in current directory if not present
</span><span class="n">sheetid</span><span class="w"> </span><span class="o">&lt;-</span><span class="s2">"1U-32UcwZP1k3saKeaH1mbvEAOfZRdNHNkWK2GI1rpPM"</span><span class="w">
</span><span class="n">gap</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gs_key</span><span class="p">(</span><span class="n">sheetid</span><span class="p">)</span><span class="w">
</span><span class="n">mysheet</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gs_read</span><span class="p">(</span><span class="n">gap</span><span class="p">,</span><span class="w"> </span><span class="n">skip</span><span class="o">=</span><span class="m">4</span><span class="p">)</span><span class="w">
</span><span class="n">myDF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">mysheet</span><span class="p">)</span><span class="w">
</span><span class="n">myDF</span><span class="w">
</span></code></pre>
</div>

<h2 id="export-of-tabular-data">Export of tabular data</h2>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">write.table</span><span class="p">(</span><span class="n">myDF</span><span class="p">,</span><span class="w"> </span><span class="n">file</span><span class="o">=</span><span class="s2">"myfile.xls"</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">,</span><span class="w"> </span><span class="n">quote</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">col.names</span><span class="o">=</span><span class="kc">NA</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<h2 id="line-wise-import">Line-wise import</h2>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">myDF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">readLines</span><span class="p">(</span><span class="s2">"myData.txt"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<h2 id="line-wise-export">Line-wise export</h2>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">writeLines</span><span class="p">(</span><span class="nb">month.name</span><span class="p">,</span><span class="w"> </span><span class="s2">"myData.txt"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<h2 id="copy-and-paste-into-r">Copy and paste into R</h2>

<p>On Windows/Linux systems</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">read.delim</span><span class="p">(</span><span class="s2">"clipboard"</span><span class="p">)</span><span class="w"> 
</span></code></pre>
</div>
<p>On Mac OS X systems</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">read.delim</span><span class="p">(</span><span class="n">pipe</span><span class="p">(</span><span class="s2">"pbpaste"</span><span class="p">))</span><span class="w"> 
</span></code></pre>
</div>

<h2 id="copy-and-paste-from-r">Copy and paste from R</h2>

<p>On Windows/Linux systems</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">write.table</span><span class="p">(</span><span class="n">iris</span><span class="p">,</span><span class="w"> </span><span class="s2">"clipboard"</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">,</span><span class="w"> </span><span class="n">col.names</span><span class="o">=</span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="n">quote</span><span class="o">=</span><span class="nb">F</span><span class="p">)</span><span class="w"> 
</span></code></pre>
</div>

<p>On Mac OS X systems</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">zz</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">pipe</span><span class="p">(</span><span class="s1">'pbcopy'</span><span class="p">,</span><span class="w"> </span><span class="s1">'w'</span><span class="p">)</span><span class="w">
</span><span class="n">write.table</span><span class="p">(</span><span class="n">iris</span><span class="p">,</span><span class="w"> </span><span class="n">zz</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">,</span><span class="w"> </span><span class="n">col.names</span><span class="o">=</span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="n">quote</span><span class="o">=</span><span class="nb">F</span><span class="p">)</span><span class="w">
</span><span class="n">close</span><span class="p">(</span><span class="n">zz</span><span class="p">)</span><span class="w"> 
</span></code></pre>
</div>

<h2 id="homework-3a">Homework 3A</h2>

<p>Homework 3A: <a href="http://girke.bioinformatics.ucr.edu/GEN242/mydoc/mydoc_homework_03.html">Object Subsetting Routines and Import/Export</a></p>

<h1 id="useful-r-functions">Useful R Functions</h1>

<h2 id="unique-entries">Unique entries</h2>

<p>Make vector entries unique with <code class="highlighter-rouge">unique</code></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">length</span><span class="p">(</span><span class="n">iris</span><span class="o">$</span><span class="n">Sepal.Length</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 150
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">iris</span><span class="o">$</span><span class="n">Sepal.Length</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 35
</code></pre>
</div>

<h2 id="count-occurrences">Count occurrences</h2>

<p>Count occurrences of entries with <code class="highlighter-rouge">table</code></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">table</span><span class="p">(</span><span class="n">iris</span><span class="o">$</span><span class="n">Species</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## 
##     setosa versicolor  virginica 
##         50         50         50
</code></pre>
</div>

<h2 id="aggregate-data">Aggregate data</h2>

<p>Compute aggregate statistics with <code class="highlighter-rouge">aggregate</code></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">aggregate</span><span class="p">(</span><span class="n">iris</span><span class="p">[,</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">],</span><span class="w"> </span><span class="n">by</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="n">iris</span><span class="o">$</span><span class="n">Species</span><span class="p">),</span><span class="w"> </span><span class="n">FUN</span><span class="o">=</span><span class="n">mean</span><span class="p">,</span><span class="w"> </span><span class="n">na.rm</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##      Group.1 Sepal.Length Sepal.Width Petal.Length Petal.Width
## 1     setosa        5.006       3.428        1.462       0.246
## 2 versicolor        5.936       2.770        4.260       1.326
## 3  virginica        6.588       2.974        5.552       2.026
</code></pre>
</div>

<h2 id="intersect-data">Intersect data</h2>

<p>Compute intersect between two vectors with <code class="highlighter-rouge">%in%</code></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nb">month.name</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"May"</span><span class="p">,</span><span class="w"> </span><span class="s2">"July"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##  [1] FALSE FALSE FALSE FALSE  TRUE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE
</code></pre>
</div>

<h2 id="merge-data-frames">Merge data frames</h2>

<p>Join two data frames by common field entries with <code class="highlighter-rouge">merge</code> (here row names <code class="highlighter-rouge">by.x=0</code>). To obtain only the common rows, change <code class="highlighter-rouge">all=TRUE</code> to <code class="highlighter-rouge">all=FALSE</code>. To merge on specific columns, refer to them by their position numbers or their column names.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">frame1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">iris</span><span class="p">[</span><span class="n">sample</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="nf">length</span><span class="p">(</span><span class="n">iris</span><span class="p">[,</span><span class="m">1</span><span class="p">]),</span><span class="w"> </span><span class="m">30</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">
</span><span class="n">frame1</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">,]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 3          4.7         3.2          1.3         0.2  setosa
## 5          5.0         3.6          1.4         0.2  setosa
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">dim</span><span class="p">(</span><span class="n">frame1</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 30  5
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">my_result</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">merge</span><span class="p">(</span><span class="n">frame1</span><span class="p">,</span><span class="w"> </span><span class="n">iris</span><span class="p">,</span><span class="w"> </span><span class="n">by.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">by.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">all</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="nf">dim</span><span class="p">(</span><span class="n">my_result</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 150  11
</code></pre>
</div>

<h1 id="sqlite-databases">SQLite Databases</h1>

<p><code class="highlighter-rouge">SQLite</code> is a lightweight relational database solution. The <code class="highlighter-rouge">RSQLite</code> package provides an easy to use interface to create, manage and query <code class="highlighter-rouge">SQLite</code> databases directly from R. Basic instructions
for using <code class="highlighter-rouge">SQLite</code> from the command-line are available <a href="https://www.sqlite.org/cli.html">here</a>. A short introduction to <code class="highlighter-rouge">RSQLite</code> is available <a href="https://github.com/rstats-db/RSQLite/blob/master/vignettes/RSQLite.Rmd">here</a>.</p>

<h2 id="loading-data-into-sqlite-databases">Loading data into SQLite databases</h2>

<p>The following loads two <code class="highlighter-rouge">data.frames</code> derived from the <code class="highlighter-rouge">iris</code> data set (here <code class="highlighter-rouge">mydf1</code> and <code class="highlighter-rouge">mydf2</code>) 
into an SQLite database (here <code class="highlighter-rouge">test.db</code>).</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">RSQLite</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Loading required package: DBI
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">mydb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">dbConnect</span><span class="p">(</span><span class="n">SQLite</span><span class="p">(),</span><span class="w"> </span><span class="s2">"test.db"</span><span class="p">)</span><span class="w"> </span><span class="c1"># Creates database file test.db
</span><span class="n">mydf1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">ids</span><span class="o">=</span><span class="n">paste0</span><span class="p">(</span><span class="s2">"id"</span><span class="p">,</span><span class="w"> </span><span class="nf">seq_along</span><span class="p">(</span><span class="n">iris</span><span class="p">[,</span><span class="m">1</span><span class="p">])),</span><span class="w"> </span><span class="n">iris</span><span class="p">)</span><span class="w">
</span><span class="n">mydf2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mydf1</span><span class="p">[</span><span class="n">sample</span><span class="p">(</span><span class="nf">seq_along</span><span class="p">(</span><span class="n">mydf1</span><span class="p">[,</span><span class="m">1</span><span class="p">]),</span><span class="w"> </span><span class="m">10</span><span class="p">),]</span><span class="w">
</span><span class="n">dbWriteTable</span><span class="p">(</span><span class="n">mydb</span><span class="p">,</span><span class="w"> </span><span class="s2">"mydf1"</span><span class="p">,</span><span class="w"> </span><span class="n">mydf1</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] TRUE
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">dbWriteTable</span><span class="p">(</span><span class="n">mydb</span><span class="p">,</span><span class="w"> </span><span class="s2">"mydf2"</span><span class="p">,</span><span class="w"> </span><span class="n">mydf2</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] TRUE
</code></pre>
</div>

<h2 id="list-names-of-tables-in-database">List names of tables in database</h2>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">dbListTables</span><span class="p">(</span><span class="n">mydb</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "mydf1" "mydf2"
</code></pre>
</div>

<h2 id="import-table-into-dataframe">Import table into <code class="highlighter-rouge">data.frame</code></h2>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">dbGetQuery</span><span class="p">(</span><span class="n">mydb</span><span class="p">,</span><span class="w"> </span><span class="s1">'SELECT * FROM mydf2'</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##      ids Sepal.Length Sepal.Width Petal.Length Petal.Width    Species
## 1  id115          5.8         2.8          5.1         2.4  virginica
## 2   id73          6.3         2.5          4.9         1.5 versicolor
## 3   id80          5.7         2.6          3.5         1.0 versicolor
## 4   id63          6.0         2.2          4.0         1.0 versicolor
## 5  id123          7.7         2.8          6.7         2.0  virginica
## 6   id99          5.1         2.5          3.0         1.1 versicolor
## 7  id126          7.2         3.2          6.0         1.8  virginica
## 8    id2          4.9         3.0          1.4         0.2     setosa
## 9   id91          5.5         2.6          4.4         1.2 versicolor
## 10  id20          5.1         3.8          1.5         0.3     setosa
</code></pre>
</div>

<h2 id="query-database">Query database</h2>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">dbGetQuery</span><span class="p">(</span><span class="n">mydb</span><span class="p">,</span><span class="w"> </span><span class="s1">'SELECT * FROM mydf1 WHERE "Sepal.Length" &lt; 4.6'</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##    ids Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 1  id9          4.4         2.9          1.4         0.2  setosa
## 2 id14          4.3         3.0          1.1         0.1  setosa
## 3 id39          4.4         3.0          1.3         0.2  setosa
## 4 id42          4.5         2.3          1.3         0.3  setosa
## 5 id43          4.4         3.2          1.3         0.2  setosa
</code></pre>
</div>

<h2 id="join-tables">Join tables</h2>

<p>The two tables can be joined on the shared <code class="highlighter-rouge">ids</code> column as follows.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">dbGetQuery</span><span class="p">(</span><span class="n">mydb</span><span class="p">,</span><span class="w"> </span><span class="s1">'SELECT * FROM mydf1, mydf2 WHERE mydf1.ids = mydf2.ids'</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##      ids Sepal.Length Sepal.Width Petal.Length Petal.Width    Species   ids Sepal.Length
## 1    id2          4.9         3.0          1.4         0.2     setosa   id2          4.9
## 2   id20          5.1         3.8          1.5         0.3     setosa  id20          5.1
## 3   id63          6.0         2.2          4.0         1.0 versicolor  id63          6.0
## 4   id73          6.3         2.5          4.9         1.5 versicolor  id73          6.3
## 5   id80          5.7         2.6          3.5         1.0 versicolor  id80          5.7
## 6   id91          5.5         2.6          4.4         1.2 versicolor  id91          5.5
## 7   id99          5.1         2.5          3.0         1.1 versicolor  id99          5.1
## 8  id115          5.8         2.8          5.1         2.4  virginica id115          5.8
## 9  id123          7.7         2.8          6.7         2.0  virginica id123          7.7
## 10 id126          7.2         3.2          6.0         1.8  virginica id126          7.2
##    Sepal.Width Petal.Length Petal.Width    Species
## 1          3.0          1.4         0.2     setosa
## 2          3.8          1.5         0.3     setosa
## 3          2.2          4.0         1.0 versicolor
## 4          2.5          4.9         1.5 versicolor
## 5          2.6          3.5         1.0 versicolor
## 6          2.6          4.4         1.2 versicolor
## 7          2.5          3.0         1.1 versicolor
## 8          2.8          5.1         2.4  virginica
## 9          2.8          6.7         2.0  virginica
## 10         3.2          6.0         1.8  virginica
</code></pre>
</div>

<h1 id="graphics-in-r">Graphics in R</h1>

<h2 id="advantages">Advantages</h2>

<ul>
  <li>Powerful environment for visualizing scientific data</li>
  <li>Integrated graphics and statistics infrastructure</li>
  <li>Publication quality graphics</li>
  <li>Fully programmable</li>
  <li>Highly reproducible</li>
  <li>Full <a href="http://www.latex-project.org/">LaTeX</a> and Markdown support via <code class="highlighter-rouge">knitr</code> and <code class="highlighter-rouge">R markdown</code></li>
  <li>Vast number of R packages with graphics utilities</li>
</ul>

<h2 id="documentation-for-r-graphics">Documentation for R Graphics</h2>

<p><strong>General</strong></p>

<ul>
  <li>Graphics Task Page - <a href="http://cran.r-project.org/web/views/Graphics.html">URL</a></li>
  <li>R Graph Gallery - <a href="http://addictedtor.free.fr/graphiques/allgraph.php">URL</a></li>
  <li>R Graphical Manual - <a href="http://cged.genes.nig.ac.jp/RGM2/index.php">URL</a></li>
  <li>Paul Murrell’s book R (Grid) Graphics - <a href="http://www.stat.auckland.ac.nz/~paul/RGraphics/rgraphics.html">URL</a></li>
</ul>

<p><strong>Interactive graphics</strong></p>

<ul>
  <li>rggobi` (GGobi) - <a href="http://www.ggobi.org/">URL</a></li>
  <li><code class="highlighter-rouge">iplots</code> - <a href="http://www.rosuda.org/iplots/">URL</a></li>
  <li>Open GL (<code class="highlighter-rouge">rgl</code>) - <a href="http://rgl.neoscientists.org/gallery.shtml">URL</a></li>
</ul>

<h2 id="graphics-environments">Graphics Environments</h2>

<p><strong>Viewing and saving graphics in R</strong></p>

<ul>
  <li>On-screen graphics</li>
  <li>postscript, pdf, svg</li>
  <li>jpeg, png, wmf, tiff, …</li>
</ul>

<p><strong>Four major graphic environments</strong></p>

<p>(a) Low-level infrastructure</p>

<ul>
  <li>R Base Graphics (low- and high-level)</li>
  <li><code class="highlighter-rouge">grid</code>: <a href="http://www.stat.auckland.ac.nz/~paul/grid/grid.html">Manual</a></li>
</ul>

<p>(b) High-level infrastructure
        \begin{itemize}
- <code class="highlighter-rouge">lattice</code>: <a href="http://lmdvr.r-forge.r-project.org">Manual</a>, <a href="http://www.his.sunderland.ac.uk/~cs0her/Statistics/UsingLatticeGraphicsInR.htm">Intro</a>, <a href="http://www.amazon.com/Lattice-Multivariate-Data-Visualization-Use/dp/0387759689">Book</a>
- <code class="highlighter-rouge">ggplot2</code>: <a href="http://had.co.nz/ggplot2/">Manual</a>, <a href="http://www.ling.upenn.edu/~joseff/rstudy/summer2010_ggplot2_intro.html">Intro</a>, <a href="http://had.co.nz/ggplot2/book/">Book</a></p>

<h2 id="base-graphics-overview">Base Graphics: Overview</h2>

<p><strong>Important high-level plotting functions</strong></p>

<ul>
  <li><code class="highlighter-rouge">plot</code>: generic x-y plotting</li>
  <li><code class="highlighter-rouge">barplot</code>: bar plots</li>
  <li><code class="highlighter-rouge">boxplot</code>: box-and-whisker plot</li>
  <li><code class="highlighter-rouge">hist</code>: histograms</li>
  <li><code class="highlighter-rouge">pie</code>: pie charts</li>
  <li><code class="highlighter-rouge">dotchart</code>: cleveland dot plots</li>
  <li><code class="highlighter-rouge">image, heatmap, contour, persp</code>: functions to generate image-like plots</li>
  <li><code class="highlighter-rouge">qqnorm, qqline, qqplot</code>: distribution comparison plots</li>
  <li><code class="highlighter-rouge">pairs, coplot</code>: display of multivariant data</li>
</ul>

<p><strong>Help on graphics functions</strong></p>

<ul>
  <li><code class="highlighter-rouge">?myfct</code></li>
  <li><code class="highlighter-rouge">?plot</code></li>
  <li><code class="highlighter-rouge">?par</code></li>
</ul>

<h3 id="preferred-object-types">Preferred Object Types</h3>

<ul>
  <li>Matrices and data frames</li>
  <li>Vectors</li>
  <li>Named vectors</li>
</ul>

<h2 id="scatter-plots">Scatter Plots</h2>

<h3 id="basic-scatter-plot">Basic Scatter Plot</h3>

<p>Sample data set for subsequent plots</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">set.seed</span><span class="p">(</span><span class="m">1410</span><span class="p">)</span><span class="w">
</span><span class="n">y</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">matrix</span><span class="p">(</span><span class="n">runif</span><span class="p">(</span><span class="m">30</span><span class="p">),</span><span class="w"> </span><span class="n">ncol</span><span class="o">=</span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">dimnames</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="nb">letters</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">],</span><span class="w"> </span><span class="nb">LETTERS</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">3</span><span class="p">]))</span><span class="w">
</span></code></pre>
</div>

<p>Plot data</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">plot</span><span class="p">(</span><span class="n">y</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="n">y</span><span class="p">[,</span><span class="m">2</span><span class="p">])</span><span class="w"> 
</span></code></pre>
</div>

<p><img src="Rbasics_files/figure-html/basic_scatter_plot-1.png" alt="" />\</p>

<h3 id="all-pairs">All pairs</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">pairs</span><span class="p">(</span><span class="n">y</span><span class="p">)</span><span class="w"> 
</span></code></pre>
</div>

<p><img src="Rbasics_files/figure-html/pairs_scatter_plot-1.png" alt="" />\</p>

<h3 id="with-labels">With labels</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">plot</span><span class="p">(</span><span class="n">y</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="n">y</span><span class="p">[,</span><span class="m">2</span><span class="p">],</span><span class="w"> </span><span class="n">pch</span><span class="o">=</span><span class="m">20</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="o">=</span><span class="s2">"red"</span><span class="p">,</span><span class="w"> </span><span class="n">main</span><span class="o">=</span><span class="s2">"Symbols and Labels"</span><span class="p">)</span><span class="w">
</span><span class="n">text</span><span class="p">(</span><span class="n">y</span><span class="p">[,</span><span class="m">1</span><span class="p">]</span><span class="m">+0.03</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="p">[,</span><span class="m">2</span><span class="p">],</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">y</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<p><img src="Rbasics_files/figure-html/labels_scatter_plot-1.png" alt="" />\</p>

<h2 id="more-examples">More examples</h2>

<p><strong>Print instead of symbols the row names</strong></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">plot</span><span class="p">(</span><span class="n">y</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="n">y</span><span class="p">[,</span><span class="m">2</span><span class="p">],</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"n"</span><span class="p">,</span><span class="w"> </span><span class="n">main</span><span class="o">=</span><span class="s2">"Plot of Labels"</span><span class="p">)</span><span class="w">
</span><span class="n">text</span><span class="p">(</span><span class="n">y</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="n">y</span><span class="p">[,</span><span class="m">2</span><span class="p">],</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">y</span><span class="p">))</span><span class="w"> 
</span></code></pre>
</div>

<p><img src="Rbasics_files/figure-html/row_scatter_plot-1.png" alt="" />\</p>

<p><strong>Usage of important plotting parameters</strong></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">grid</span><span class="p">(</span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="n">lwd</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w"> 
</span><span class="n">op</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">par</span><span class="p">(</span><span class="n">mar</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">8</span><span class="p">,</span><span class="m">8</span><span class="p">,</span><span class="m">8</span><span class="p">,</span><span class="m">8</span><span class="p">),</span><span class="w"> </span><span class="n">bg</span><span class="o">=</span><span class="s2">"lightblue"</span><span class="p">)</span><span class="w">
</span><span class="n">plot</span><span class="p">(</span><span class="n">y</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="n">y</span><span class="p">[,</span><span class="m">2</span><span class="p">],</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"p"</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="o">=</span><span class="s2">"red"</span><span class="p">,</span><span class="w"> </span><span class="n">cex.lab</span><span class="o">=</span><span class="m">1.2</span><span class="p">,</span><span class="w"> </span><span class="n">cex.axis</span><span class="o">=</span><span class="m">1.2</span><span class="p">,</span><span class="w"> 
     </span><span class="n">cex.main</span><span class="o">=</span><span class="m">1.2</span><span class="p">,</span><span class="w"> </span><span class="n">cex.sub</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">lwd</span><span class="o">=</span><span class="m">4</span><span class="p">,</span><span class="w"> </span><span class="n">pch</span><span class="o">=</span><span class="m">20</span><span class="p">,</span><span class="w"> </span><span class="n">xlab</span><span class="o">=</span><span class="s2">"x label"</span><span class="p">,</span><span class="w"> 
     </span><span class="n">ylab</span><span class="o">=</span><span class="s2">"y label"</span><span class="p">,</span><span class="w"> </span><span class="n">main</span><span class="o">=</span><span class="s2">"My Main"</span><span class="p">,</span><span class="w"> </span><span class="n">sub</span><span class="o">=</span><span class="s2">"My Sub"</span><span class="p">)</span><span class="w">
</span><span class="n">par</span><span class="p">(</span><span class="n">op</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>
<p><em>_Important arguments</em></p>

<ul>
  <li><code class="highlighter-rouge">mar</code>: specifies the margin sizes around the plotting area in order: <code class="highlighter-rouge">c(bottom, left, top, right)</code></li>
  <li><code class="highlighter-rouge">col</code>: color of symbols</li>
  <li><code class="highlighter-rouge">pch</code>: type of symbols, samples: <code class="highlighter-rouge">example(points)</code></li>
  <li><code class="highlighter-rouge">lwd</code>: size of symbols</li>
  <li><code class="highlighter-rouge">cex.*</code>: control font sizes</li>
  <li>For details see <code class="highlighter-rouge">?par</code></li>
</ul>

<h3 id="add-regression-line">Add regression line</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">plot</span><span class="p">(</span><span class="n">y</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="n">y</span><span class="p">[,</span><span class="m">2</span><span class="p">])</span><span class="w">
</span><span class="n">myline</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">lm</span><span class="p">(</span><span class="n">y</span><span class="p">[,</span><span class="m">2</span><span class="p">]</span><span class="o">~</span><span class="n">y</span><span class="p">[,</span><span class="m">1</span><span class="p">]);</span><span class="w"> </span><span class="n">abline</span><span class="p">(</span><span class="n">myline</span><span class="p">,</span><span class="w"> </span><span class="n">lwd</span><span class="o">=</span><span class="m">2</span><span class="p">)</span><span class="w"> 
</span></code></pre>
</div>

<p><img src="Rbasics_files/figure-html/plot_regression-1.png" alt="" />\</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">summary</span><span class="p">(</span><span class="n">myline</span><span class="p">)</span><span class="w"> 
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## 
## Call:
## lm(formula = y[, 2] ~ y[, 1])
## 
## Residuals:
##      Min       1Q   Median       3Q      Max 
## -0.40357 -0.17912 -0.04299  0.22147  0.46623 
## 
## Coefficients:
##             Estimate Std. Error t value Pr(&gt;|t|)  
## (Intercept)   0.5764     0.2110   2.732   0.0258 *
## y[, 1]       -0.3647     0.3959  -0.921   0.3839  
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
## 
## Residual standard error: 0.3095 on 8 degrees of freedom
## Multiple R-squared:  0.09589,	Adjusted R-squared:  -0.01712 
## F-statistic: 0.8485 on 1 and 8 DF,  p-value: 0.3839
</code></pre>
</div>

<h3 id="log-scale">Log scale</h3>

<p>Same plot as above, but on log scale</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">plot</span><span class="p">(</span><span class="n">y</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="n">y</span><span class="p">[,</span><span class="m">2</span><span class="p">],</span><span class="w"> </span><span class="n">log</span><span class="o">=</span><span class="s2">"xy"</span><span class="p">)</span><span class="w"> 
</span></code></pre>
</div>

<p><img src="Rbasics_files/figure-html/plot_regression_log-1.png" alt="" />\</p>

<h3 id="add-a-mathematical-expression">Add a mathematical expression</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">plot</span><span class="p">(</span><span class="n">y</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="n">y</span><span class="p">[,</span><span class="m">2</span><span class="p">]);</span><span class="w"> </span><span class="n">text</span><span class="p">(</span><span class="n">y</span><span class="p">[</span><span class="m">1</span><span class="p">,</span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="n">y</span><span class="p">[</span><span class="m">1</span><span class="p">,</span><span class="m">2</span><span class="p">],</span><span class="w"> </span><span class="nf">expression</span><span class="p">(</span><span class="nf">sum</span><span class="p">(</span><span class="n">frac</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="nf">sqrt</span><span class="p">(</span><span class="n">x</span><span class="o">^</span><span class="m">2</span><span class="o">*</span><span class="nb">pi</span><span class="p">)))),</span><span class="w"> </span><span class="n">cex</span><span class="o">=</span><span class="m">1.3</span><span class="p">)</span><span class="w"> 
</span></code></pre>
</div>

<p><img src="Rbasics_files/figure-html/plot_regression_math-1.png" alt="" />\</p>

<h2 id="homework-3b">Homework 3B</h2>

<p>Homework 3B: <a href="http://girke.bioinformatics.ucr.edu/GEN242/mydoc/mydoc_homework_03.html">Scatter Plots</a></p>

<h2 id="line-plots">Line Plots</h2>

<h3 id="single-data-set">Single data set</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">plot</span><span class="p">(</span><span class="n">y</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"l"</span><span class="p">,</span><span class="w"> </span><span class="n">lwd</span><span class="o">=</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="o">=</span><span class="s2">"blue"</span><span class="p">)</span><span class="w"> 
</span></code></pre>
</div>

<p><img src="Rbasics_files/figure-html/plot_line_single-1.png" alt="" />\</p>

<h3 id="many-data-sets">Many Data Sets</h3>

<p>Plots line graph for all columns in data frame <code class="highlighter-rouge">y</code>. The <code class="highlighter-rouge">split.screen</code> function is used in this example in a for loop to overlay several line graphs in the same plot.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">split.screen</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">1</span><span class="p">))</span><span class="w"> 
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 1
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">plot</span><span class="p">(</span><span class="n">y</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">1</span><span class="p">),</span><span class="w"> </span><span class="n">xlab</span><span class="o">=</span><span class="s2">"Measurement"</span><span class="p">,</span><span class="w"> </span><span class="n">ylab</span><span class="o">=</span><span class="s2">"Intensity"</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"l"</span><span class="p">,</span><span class="w"> </span><span class="n">lwd</span><span class="o">=</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="o">=</span><span class="m">1</span><span class="p">)</span><span class="w">
</span><span class="k">for</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">2</span><span class="o">:</span><span class="nf">length</span><span class="p">(</span><span class="n">y</span><span class="p">[</span><span class="m">1</span><span class="p">,]))</span><span class="w"> </span><span class="p">{</span><span class="w"> 
	</span><span class="n">screen</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">new</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span><span class="w">
	</span><span class="n">plot</span><span class="p">(</span><span class="n">y</span><span class="p">[,</span><span class="n">i</span><span class="p">],</span><span class="w"> </span><span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">1</span><span class="p">),</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"l"</span><span class="p">,</span><span class="w"> </span><span class="n">lwd</span><span class="o">=</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="o">=</span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="n">xaxt</span><span class="o">=</span><span class="s2">"n"</span><span class="p">,</span><span class="w"> </span><span class="n">yaxt</span><span class="o">=</span><span class="s2">"n"</span><span class="p">,</span><span class="w"> </span><span class="n">ylab</span><span class="o">=</span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">xlab</span><span class="o">=</span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">main</span><span class="o">=</span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">bty</span><span class="o">=</span><span class="s2">"n"</span><span class="p">)</span><span class="w"> 
</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>

<p><img src="Rbasics_files/figure-html/plot_line_many-1.png" alt="" />\</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">close.screen</span><span class="p">(</span><span class="n">all</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w"> 
</span></code></pre>
</div>

<h2 id="bar-plots">Bar Plots</h2>

<h3 id="basics">Basics</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">barplot</span><span class="p">(</span><span class="n">y</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">,],</span><span class="w"> </span><span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="nf">max</span><span class="p">(</span><span class="n">y</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">,])</span><span class="m">+0.3</span><span class="p">),</span><span class="w"> </span><span class="n">beside</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">legend</span><span class="o">=</span><span class="nb">letters</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">])</span><span class="w"> 
</span><span class="n">text</span><span class="p">(</span><span class="n">labels</span><span class="o">=</span><span class="nf">round</span><span class="p">(</span><span class="n">as.vector</span><span class="p">(</span><span class="n">as.matrix</span><span class="p">(</span><span class="n">y</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">,])),</span><span class="m">2</span><span class="p">),</span><span class="w"> </span><span class="n">x</span><span class="o">=</span><span class="n">seq</span><span class="p">(</span><span class="m">1.5</span><span class="p">,</span><span class="w"> </span><span class="m">13</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="o">=</span><span class="m">1</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">sort</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">1</span><span class="p">,</span><span class="m">2</span><span class="p">),</span><span class="w"> </span><span class="m">4</span><span class="p">)),</span><span class="w"> </span><span class="n">y</span><span class="o">=</span><span class="n">as.vector</span><span class="p">(</span><span class="n">as.matrix</span><span class="p">(</span><span class="n">y</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">,]))</span><span class="m">+0.04</span><span class="p">)</span><span class="w"> 
</span></code></pre>
</div>

<p><img src="Rbasics_files/figure-html/plot_bar_simple-1.png" alt="" />\</p>

<h3 id="error-bars">Error Bars</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">bar</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">barplot</span><span class="p">(</span><span class="n">m</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rowMeans</span><span class="p">(</span><span class="n">y</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">10</span><span class="p">))</span><span class="w">
</span><span class="n">stdev</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sd</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="n">y</span><span class="p">))</span><span class="w">
</span><span class="n">arrows</span><span class="p">(</span><span class="n">bar</span><span class="p">,</span><span class="w"> </span><span class="n">m</span><span class="p">,</span><span class="w"> </span><span class="n">bar</span><span class="p">,</span><span class="w"> </span><span class="n">m</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">stdev</span><span class="p">,</span><span class="w"> </span><span class="n">length</span><span class="o">=</span><span class="m">0.15</span><span class="p">,</span><span class="w"> </span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">90</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="Rbasics_files/figure-html/plot_bar_error-1.png" alt="" />\</p>

<h2 id="histograms">Histograms</h2>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">hist</span><span class="p">(</span><span class="n">y</span><span class="p">,</span><span class="w"> </span><span class="n">freq</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">breaks</span><span class="o">=</span><span class="m">10</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="Rbasics_files/figure-html/plot_hist-1.png" alt="" />\</p>

<h2 id="density-plots">Density Plots</h2>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">plot</span><span class="p">(</span><span class="n">density</span><span class="p">(</span><span class="n">y</span><span class="p">),</span><span class="w"> </span><span class="n">col</span><span class="o">=</span><span class="s2">"red"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="Rbasics_files/figure-html/plot_dens-1.png" alt="" />\</p>

<h2 id="pie-charts">Pie Charts</h2>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">pie</span><span class="p">(</span><span class="n">y</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="n">col</span><span class="o">=</span><span class="n">rainbow</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">y</span><span class="p">[,</span><span class="m">1</span><span class="p">]),</span><span class="w"> </span><span class="n">start</span><span class="o">=</span><span class="m">0.1</span><span class="p">,</span><span class="w"> </span><span class="n">end</span><span class="o">=</span><span class="m">0.8</span><span class="p">),</span><span class="w"> </span><span class="n">clockwise</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">legend</span><span class="p">(</span><span class="s2">"topright"</span><span class="p">,</span><span class="w"> </span><span class="n">legend</span><span class="o">=</span><span class="n">row.names</span><span class="p">(</span><span class="n">y</span><span class="p">),</span><span class="w"> </span><span class="n">cex</span><span class="o">=</span><span class="m">1.3</span><span class="p">,</span><span class="w"> </span><span class="n">bty</span><span class="o">=</span><span class="s2">"n"</span><span class="p">,</span><span class="w"> </span><span class="n">pch</span><span class="o">=</span><span class="m">15</span><span class="p">,</span><span class="w"> </span><span class="n">pt.cex</span><span class="o">=</span><span class="m">1.8</span><span class="p">,</span><span class="w"> 
</span><span class="n">col</span><span class="o">=</span><span class="n">rainbow</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">y</span><span class="p">[,</span><span class="m">1</span><span class="p">]),</span><span class="w"> </span><span class="n">start</span><span class="o">=</span><span class="m">0.1</span><span class="p">,</span><span class="w"> </span><span class="n">end</span><span class="o">=</span><span class="m">0.8</span><span class="p">),</span><span class="w"> </span><span class="n">ncol</span><span class="o">=</span><span class="m">1</span><span class="p">)</span><span class="w"> 
</span></code></pre>
</div>

<p><img src="Rbasics_files/figure-html/plot_pie-1.png" alt="" />\</p>

<h2 id="color-selection-utilities">Color Selection Utilities</h2>

<p>Default color palette and how to change it</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">palette</span><span class="p">()</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "black"   "red"     "green3"  "blue"    "cyan"    "magenta" "yellow"  "gray"
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">palette</span><span class="p">(</span><span class="n">rainbow</span><span class="p">(</span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="n">start</span><span class="o">=</span><span class="m">0.1</span><span class="p">,</span><span class="w"> </span><span class="n">end</span><span class="o">=</span><span class="m">0.2</span><span class="p">))</span><span class="w">
</span><span class="n">palette</span><span class="p">()</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "#FF9900" "#FFBF00" "#FFE600" "#F2FF00" "#CCFF00"
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">palette</span><span class="p">(</span><span class="s2">"default"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p>The <code class="highlighter-rouge">gray</code> function allows to select any type of gray shades by providing values from 0 to 1</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">gray</span><span class="p">(</span><span class="n">seq</span><span class="p">(</span><span class="m">0.1</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="o">=</span><span class="w"> </span><span class="m">0.2</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "#1A1A1A" "#4D4D4D" "#808080" "#B3B3B3" "#E6E6E6"
</code></pre>
</div>

<p>Color gradients with <code class="highlighter-rouge">colorpanel</code> function from <code class="highlighter-rouge">gplots</code> library`</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">gplots</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## 
## Attaching package: 'gplots'
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## The following object is masked from 'package:stats':
## 
##     lowess
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">colorpanel</span><span class="p">(</span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="s2">"darkblue"</span><span class="p">,</span><span class="w"> </span><span class="s2">"yellow"</span><span class="p">,</span><span class="w"> </span><span class="s2">"white"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "#00008B" "#808046" "#FFFF00" "#FFFF80" "#FFFFFF"
</code></pre>
</div>
<p>Much more on colors in R see Earl Glynn’s color chart <a href="http://research.stowers-institute.org/efg/R/Color/Chart/">here</a></p>

<h2 id="saving-graphics-to-file">Saving Graphics to File</h2>

<p>After the <code class="highlighter-rouge">pdf()</code> command all graphs are redirected to file <code class="highlighter-rouge">test.pdf</code>. Works for all common formats similarly: jpeg, png, ps, tiff, …</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">pdf</span><span class="p">(</span><span class="s2">"test.pdf"</span><span class="p">)</span><span class="w">
</span><span class="n">plot</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">)</span><span class="w">
</span><span class="n">dev.off</span><span class="p">()</span><span class="w"> 
</span></code></pre>
</div>

<p>Generates Scalable Vector Graphics (SVG) files that can be edited in vector graphics programs, such as InkScape.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="s2">"RSvgDevice"</span><span class="p">)</span><span class="w">
</span><span class="n">devSVG</span><span class="p">(</span><span class="s2">"test.svg"</span><span class="p">)</span><span class="w">
</span><span class="n">plot</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">)</span><span class="w">
</span><span class="n">dev.off</span><span class="p">()</span><span class="w"> 
</span></code></pre>
</div>

<h2 id="homework-3c">Homework 3C</h2>

<p>Homework 3C: <a href="http://girke.bioinformatics.ucr.edu/GEN242/mydoc/mydoc_homework_03.html">Bar Plots</a></p>

<h1 id="analysis-routine">Analysis Routine</h1>

<h2 id="overview-1">Overview</h2>

<p>The following exercise introduces a variety of useful data analysis utilities in R.</p>

<h2 id="analysis-routine-data-import">Analysis Routine: Data Import</h2>

<ul>
  <li>
    <p><strong>Step 1</strong>: To get started with this exercise, direct your R session to a dedicated workshop directory and download into this directory the following sample tables. Then import the files into Excel and save them as tab delimited text files.</p>

    <ul>
      <li><a href="http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/Samples/MolecularWeight_tair7.xls">MolecularWeight_tair7.xls</a></li>
      <li><a href="http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/Samples/TargetP_analysis_tair7.xls">TargetP_analysis_tair7.xls</a></li>
    </ul>
  </li>
</ul>

<p><strong>Import the tables into R</strong></p>

<p>Import molecular weight table</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">my_mw</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.delim</span><span class="p">(</span><span class="n">file</span><span class="o">=</span><span class="s2">"MolecularWeight_tair7.xls"</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="o">=</span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">)</span><span class="w"> 
</span><span class="n">my_mw</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">,]</span><span class="w">
</span></code></pre>
</div>

<p>Import subcelluar targeting table</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">my_target</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.delim</span><span class="p">(</span><span class="n">file</span><span class="o">=</span><span class="s2">"TargetP_analysis_tair7.xls"</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="o">=</span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">)</span><span class="w"> 
</span><span class="n">my_target</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">,]</span><span class="w">
</span></code></pre>
</div>

<p>Online import of molecular weight table</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">my_mw</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.delim</span><span class="p">(</span><span class="n">file</span><span class="o">=</span><span class="s2">"http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/Samples/MolecularWeight_tair7.xls"</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="o">=</span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">)</span><span class="w"> 
</span><span class="n">my_mw</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">,]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   Sequence.id Molecular.Weight.Da. Residues
## 1 AT1G08520.1                83285      760
## 2 AT1G08530.1                27015      257
</code></pre>
</div>

<p>Online import of subcelluar targeting table</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">my_target</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.delim</span><span class="p">(</span><span class="n">file</span><span class="o">=</span><span class="s2">"http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/Samples/TargetP_analysis_tair7.xls"</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="o">=</span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">)</span><span class="w"> 
</span><span class="n">my_target</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">,]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##      GeneName Loc   cTP   mTP    SP other
## 1 AT1G08520.1   C 0.822 0.137 0.029 0.039
## 2 AT1G08530.1   C 0.817 0.058 0.010 0.100
</code></pre>
</div>

<h2 id="merging-data-frames">Merging Data Frames</h2>

<ul>
  <li><strong>Step 2</strong>: Assign uniform gene ID column titles</li>
</ul>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">colnames</span><span class="p">(</span><span class="n">my_target</span><span class="p">)[</span><span class="m">1</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"ID"</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">my_mw</span><span class="p">)[</span><span class="m">1</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"ID"</span><span class="w"> 
</span></code></pre>
</div>

<ul>
  <li><strong>Step 3</strong>: Merge the two tables based on common ID field</li>
</ul>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">my_mw_target</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">merge</span><span class="p">(</span><span class="n">my_mw</span><span class="p">,</span><span class="w"> </span><span class="n">my_target</span><span class="p">,</span><span class="w"> </span><span class="n">by.x</span><span class="o">=</span><span class="s2">"ID"</span><span class="p">,</span><span class="w"> </span><span class="n">by.y</span><span class="o">=</span><span class="s2">"ID"</span><span class="p">,</span><span class="w"> </span><span class="n">all.x</span><span class="o">=</span><span class="nb">T</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<ul>
  <li><strong>Step 4</strong>: Shorten one table before the merge and then remove the non-matching rows (NAs) in the merged file</li>
</ul>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">my_mw_target2a</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">merge</span><span class="p">(</span><span class="n">my_mw</span><span class="p">,</span><span class="w"> </span><span class="n">my_target</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">40</span><span class="p">,],</span><span class="w"> </span><span class="n">by.x</span><span class="o">=</span><span class="s2">"ID"</span><span class="p">,</span><span class="w"> </span><span class="n">by.y</span><span class="o">=</span><span class="s2">"ID"</span><span class="p">,</span><span class="w"> </span><span class="n">all.x</span><span class="o">=</span><span class="nb">T</span><span class="p">)</span><span class="w">  </span><span class="c1"># To remove non-matching rows, use the argument setting 'all=F'.
</span><span class="n">my_mw_target2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">na.omit</span><span class="p">(</span><span class="n">my_mw_target2a</span><span class="p">)</span><span class="w"> </span><span class="c1"># Removes rows containing "NAs" (non-matching rows).
</span></code></pre>
</div>

<ul>
  <li><strong>Homework 3D</strong>: How can the merge function in the previous step be executed so that only the common rows among the two data frames are returned? Prove that both methods - the two step version with <code class="highlighter-rouge">na.omit</code> and your method - return identical results.</li>
  <li><strong>Homework 3E</strong>: Replace all <code class="highlighter-rouge">NAs</code> in the data frame <code class="highlighter-rouge">my_mw_target2a</code> with zeros.</li>
</ul>

<h2 id="filtering-data">Filtering Data</h2>

<ul>
  <li><strong>Step 5</strong>: Retrieve all records with a value of greater than 100,000 in ‘MW’ column and ‘C’ value in ‘Loc’ column (targeted to chloroplast).</li>
</ul>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">query</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">my_mw_target</span><span class="p">[</span><span class="n">my_mw_target</span><span class="p">[,</span><span class="w"> </span><span class="m">2</span><span class="p">]</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">100000</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">my_mw_target</span><span class="p">[,</span><span class="w"> </span><span class="m">4</span><span class="p">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"C"</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w"> 
</span><span class="n">query</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##              ID Molecular.Weight.Da. Residues Loc   cTP   mTP    SP other
## 219 AT1G02730.1               132588     1181   C 0.972 0.038 0.008 0.045
## 243 AT1G02890.1               136825     1252   C 0.748 0.529 0.011 0.013
## 281 AT1G03160.1               100732      912   C 0.871 0.235 0.011 0.007
## 547 AT1G05380.1               126360     1138   C 0.740 0.099 0.016 0.358
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">dim</span><span class="p">(</span><span class="n">query</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 170   8
</code></pre>
</div>

<ul>
  <li><strong>Homework 3F</strong>: How many protein entries in the <code class="highlighter-rouge">my</code>_mw<code class="highlighter-rouge">_target</code> data frame have a MW of greater then 4,000 and less then 5,000. Subset the data frame accordingly and sort it by MW to check that your result is correct.</li>
</ul>

<h2 id="string-substitutions">String Substitutions</h2>

<ul>
  <li><strong>Step 6</strong>: Use a regular expression in a substitute function to generate a separate ID column that lacks the gene model extensions.
«label=Exercise 4.7, eval=TRUE, echo=TRUE, keep.source=TRUE»=</li>
</ul>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">my_mw_target3</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">loci</span><span class="o">=</span><span class="n">gsub</span><span class="p">(</span><span class="s2">"\\..*"</span><span class="p">,</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">my_mw_target</span><span class="p">[,</span><span class="m">1</span><span class="p">]),</span><span class="w"> </span><span class="n">perl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="n">my_mw_target</span><span class="p">)</span><span class="w">
</span><span class="n">my_mw_target3</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">3</span><span class="p">,</span><span class="m">1</span><span class="o">:</span><span class="m">8</span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##        loci          ID Molecular.Weight.Da. Residues Loc  cTP   mTP    SP
## 1 AT1G01010 AT1G01010.1                49426      429   _ 0.10 0.090 0.075
## 2 AT1G01020 AT1G01020.1                28092      245   * 0.01 0.636 0.158
## 3 AT1G01020 AT1G01020.2                21711      191   * 0.01 0.636 0.158
</code></pre>
</div>

<ul>
  <li><strong>Homework 3G</strong>: Retrieve those rows in <code class="highlighter-rouge">my_mw_target3</code> where the second column contains the following identifiers: <code class="highlighter-rouge">c("AT5G52930.1", "AT4G18950.1", "AT1G15385.1", "AT4G36500.1", "AT1G67530.1")</code>. Use the <code class="highlighter-rouge">%in%</code> function for this query. As an alternative approach, assign the second column to the row index of the data frame and then perform the same query again using the row index. Explain the difference of the two methods.</li>
</ul>

<h2 id="calculations-on-data-frames">Calculations on Data Frames</h2>

<ul>
  <li><strong>Step 7</strong>: Count the number of duplicates in the loci column with the <code class="highlighter-rouge">table</code> function and append the result to the data frame with the <code class="highlighter-rouge">cbind</code> function.</li>
</ul>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">mycounts</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">table</span><span class="p">(</span><span class="n">my_mw_target3</span><span class="p">[,</span><span class="m">1</span><span class="p">])[</span><span class="n">my_mw_target3</span><span class="p">[,</span><span class="m">1</span><span class="p">]]</span><span class="w">
</span><span class="n">my_mw_target4</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">cbind</span><span class="p">(</span><span class="n">my_mw_target3</span><span class="p">,</span><span class="w"> </span><span class="n">Freq</span><span class="o">=</span><span class="n">mycounts</span><span class="p">[</span><span class="nf">as.character</span><span class="p">(</span><span class="n">my_mw_target3</span><span class="p">[,</span><span class="m">1</span><span class="p">])])</span><span class="w"> 
</span></code></pre>
</div>

<ul>
  <li><strong>Step 8</strong>: Perform a vectorized devision of columns 3 and 4 (average AA weight per protein)</li>
</ul>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">data.frame</span><span class="p">(</span><span class="n">my_mw_target4</span><span class="p">,</span><span class="w"> </span><span class="n">avg_AA_WT</span><span class="o">=</span><span class="p">(</span><span class="n">my_mw_target4</span><span class="p">[,</span><span class="m">3</span><span class="p">]</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">my_mw_target4</span><span class="p">[,</span><span class="m">4</span><span class="p">]))[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">,</span><span class="m">5</span><span class="o">:</span><span class="m">11</span><span class="p">]</span><span class="w"> 
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   Loc  cTP   mTP    SP other Freq avg_AA_WT
## 1   _ 0.10 0.090 0.075 0.925    1  115.2121
## 2   * 0.01 0.636 0.158 0.448    2  114.6612
</code></pre>
</div>

<ul>
  <li><strong>Step 9</strong>: Calculate for each row the mean and standard deviation across several columns</li>
</ul>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">mymean</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">apply</span><span class="p">(</span><span class="n">my_mw_target4</span><span class="p">[,</span><span class="m">6</span><span class="o">:</span><span class="m">9</span><span class="p">],</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">mean</span><span class="p">)</span><span class="w">
</span><span class="n">mystdev</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">apply</span><span class="p">(</span><span class="n">my_mw_target4</span><span class="p">[,</span><span class="m">6</span><span class="o">:</span><span class="m">9</span><span class="p">],</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">sd</span><span class="p">,</span><span class="w"> </span><span class="n">na.rm</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">data.frame</span><span class="p">(</span><span class="n">my_mw_target4</span><span class="p">,</span><span class="w"> </span><span class="n">mean</span><span class="o">=</span><span class="n">mymean</span><span class="p">,</span><span class="w"> </span><span class="n">stdev</span><span class="o">=</span><span class="n">mystdev</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">,</span><span class="m">5</span><span class="o">:</span><span class="m">12</span><span class="p">]</span><span class="w"> 
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   Loc  cTP   mTP    SP other Freq   mean     stdev
## 1   _ 0.10 0.090 0.075 0.925    1 0.2975 0.4184595
## 2   * 0.01 0.636 0.158 0.448    2 0.3130 0.2818912
</code></pre>
</div>

<h2 id="plotting-example">Plotting Example</h2>

<ul>
  <li><strong>Step 10</strong>: Generate scatter plot columns: ‘MW’ and ‘Residues’</li>
</ul>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">plot</span><span class="p">(</span><span class="n">my_mw_target4</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">500</span><span class="p">,</span><span class="m">3</span><span class="o">:</span><span class="m">4</span><span class="p">],</span><span class="w"> </span><span class="n">col</span><span class="o">=</span><span class="s2">"red"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="Rbasics_files/figure-html/plot_example-1.png" alt="" />\</p>

<h2 id="export-results-and-run-entire-exercise-as-script">Export Results and Run Entire Exercise as Script</h2>

<ul>
  <li><strong>Step 11</strong>: Write the data frame <code class="highlighter-rouge">my_mw_target4</code> into a tab-delimited text file and inspect it in Excel.</li>
</ul>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">write.table</span><span class="p">(</span><span class="n">my_mw_target4</span><span class="p">,</span><span class="w"> </span><span class="n">file</span><span class="o">=</span><span class="s2">"my_file.xls"</span><span class="p">,</span><span class="w"> </span><span class="n">quote</span><span class="o">=</span><span class="nb">F</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">,</span><span class="w"> </span><span class="n">col.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span><span class="p">)</span><span class="w"> 
</span></code></pre>
</div>

<ul>
  <li><strong>Homework 3H</strong>: Write all commands from this exercise into an R script named <code class="highlighter-rouge">exerciseRbasics.R</code>, or download it from <a href="http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/My_R_Scripts/exerciseRbasics.R">here</a>. Then execute the script with the <code class="highlighter-rouge">source</code> function like this: <code class="highlighter-rouge">source("exerciseRbasics.R")</code>. This will run all commands of this exercise and generate the corresponding output files in the current working directory.</li>
</ul>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">source</span><span class="p">(</span><span class="s2">"exerciseRbasics.R"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<h1 id="r-markdown">R Markdown</h1>

<h2 id="overview-2">Overview</h2>

<p>R Markdown combines markdown (an easy to write plain text format) with embedded
R code chunks. When compiling R Markdown documents, the code components can be
evaluated so that both the code and its output can be included in the final
document. This makes analysis reports highly reproducible by allowing to automatically
regenerate them when the underlying R code or data changes. R Markdown
documents (<code class="highlighter-rouge">.Rmd</code> files) can be rendered to various formats including HTML and
PDF. The R code in an <code class="highlighter-rouge">.Rmd</code> document is processed by <code class="highlighter-rouge">knitr</code>, while the
resulting <code class="highlighter-rouge">.md</code> file is rendered by <code class="highlighter-rouge">pandoc</code> to the final output formats
(<em>e.g.</em> HTML or PDF). Historically, R Markdown is an extension of the older
<code class="highlighter-rouge">Sweave/Latex</code> environment. Rendering of mathematical expressions and reference
management is also supported by R Markdown using embedded Latex syntax and
Bibtex, respectively.</p>

<h2 id="quick-start">Quick Start</h2>

<h3 id="install-r-markdown">Install R Markdown</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">install.packages</span><span class="p">(</span><span class="s2">"rmarkdown"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<h3 id="initialize-a-new-r-markdown-rmd-script">Initialize a new R Markdown (<code class="highlighter-rouge">Rmd</code>) script</h3>

<p>To minimize typing, it can be helful to start with an R Markdown template and
then modify it as needed. Note the file name of an R Markdown scirpt needs to
have the extension <code class="highlighter-rouge">.Rmd</code>. Template files for the following examples are available 
here:</p>

<ul>
  <li>R Markdown sample script: <a href="https://raw.githubusercontent.com/tgirke/GEN242/master/vignettes/07_Rbasics/sample.Rmd"><code class="highlighter-rouge">sample.Rmd</code></a></li>
  <li>Bibtex file for handling citations and reference section: <a href="https://raw.githubusercontent.com/tgirke/GEN242/master/vignettes/07_Rbasics/bibtex.bib"><code class="highlighter-rouge">bibtex.bib</code></a></li>
</ul>

<p>Users want to download these files, open the <code class="highlighter-rouge">sample.Rmd</code> file with their preferred R IDE 
(<em>e.g.</em> RStudio, vim or emacs), initilize an R session and then direct their R session to 
the location of these two files.</p>

<h3 id="metadata-section">Metadata section</h3>

<p>The metadata section (YAML header) in an R Markdown script defines how it will be processed and 
rendered. The metadata section also includes both title, author, and date information as well as 
options for customizing the output format. For instance, PDF and HTML output can be defined 
with <code class="highlighter-rouge">pdf_document</code> and <code class="highlighter-rouge">html_document</code>, respectively. The <code class="highlighter-rouge">BiocStyle::</code> prefix will use the
formatting style of the <a href="http://bioconductor.org/packages/release/bioc/html/BiocStyle.html"><code class="highlighter-rouge">BiocStyle</code></a> 
package from Bioconductor.</p>

<div class="highlighter-rouge"><pre class="highlight"><code> ---
title: "My First R Markdown Document"
author: "Author: First Last"
date: "Last update: 21 June, 2016"
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 3
    fig_caption: yes

fontsize: 14pt
bibliography: bibtex.bib
 ---
</code></pre>
</div>

<h3 id="render-rmd-script">Render <code class="highlighter-rouge">Rmd</code> script</h3>

<p>An R Markdown script can be evaluated and rendered with the following <code class="highlighter-rouge">render</code> command or by pressing the <code class="highlighter-rouge">knit</code> button in RStudio.
The <code class="highlighter-rouge">output_format</code> argument defines the format of the output (<em>e.g.</em> <code class="highlighter-rouge">html_document</code>). The setting <code class="highlighter-rouge">output_format="all"</code> will generate 
all supported output formats. Alternatively, one can specify several output formats in the metadata section as shown in the above example.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">rmarkdown</span><span class="o">::</span><span class="n">render</span><span class="p">(</span><span class="s2">"sample.Rmd"</span><span class="p">,</span><span class="w"> </span><span class="n">clean</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">output_format</span><span class="o">=</span><span class="s2">"html_document"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p>The following shows two options how to run the rendering from the command-line.</p>

<div class="language-sh highlighter-rouge"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">echo</span> <span class="s2">"rmarkdown::render('sample.Rmd', clean=TRUE)"</span> | R --slave
<span class="gp">$ </span>Rscript -e <span class="s2">"rmarkdown::render('sample.Rmd', clean=TRUE)"</span>
</code></pre>
</div>

<p>Alternatively, one can use a Makefile to evaluate and render an R Markdown
script. A sample Makefile for rendering the above <code class="highlighter-rouge">sample.Rmd</code> can be
downloaded <a href="https://raw.githubusercontent.com/tgirke/GEN242/master/vignettes/07_Rbasics/Makefile"><code class="highlighter-rouge">here</code></a>.
To apply it to a custom <code class="highlighter-rouge">Rmd</code> file, one needs open the Makefile in a text
editor and change the value assigned to <code class="highlighter-rouge">MAIN</code> (line 13) to the base name of
the corresponding <code class="highlighter-rouge">.Rmd</code> file (<em>e.g.</em> assign <code class="highlighter-rouge">systemPipeRNAseq</code> if the file
name is <code class="highlighter-rouge">systemPipeRNAseq.Rmd</code>).  To execute the <code class="highlighter-rouge">Makefile</code>, run the following
command from the command-line.</p>

<div class="language-sh highlighter-rouge"><pre class="highlight"><code><span class="gp">$ </span>make -B
</code></pre>
</div>

<h3 id="r-code-chunks">R code chunks</h3>

<p>R Code Chunks can be embedded in an R Markdown script by using three backticks
at the beginning of a new line along with arguments enclosed in curly braces
controlling the behavior of the code. The following lines contain the
plain R code. A code chunk is terminated by a new line starting with three backticks.
The following shows an example of such a code chunk. Note the backslashes are
not part of it. They have been added to print the code chunk syntax in this document.</p>

<div class="highlighter-rouge"><pre class="highlight"><code>	```\{r code_chunk_name, eval=FALSE\}
	x &lt;- 1:10
	```
</code></pre>
</div>

<p>The following lists the most important arguments to control the behavior of R code chunks:</p>

<ul>
  <li><code class="highlighter-rouge">r</code>: specifies language for code chunk, here R</li>
  <li><code class="highlighter-rouge">chode_chunk_name</code>: name of code chunk; this name needs to be unique</li>
  <li><code class="highlighter-rouge">eval</code>: if assigned <code class="highlighter-rouge">TRUE</code> the code will be evaluated</li>
  <li><code class="highlighter-rouge">warning</code>: if assigned <code class="highlighter-rouge">FALSE</code> warnings will not be shown</li>
  <li><code class="highlighter-rouge">message</code>: if assigned <code class="highlighter-rouge">FALSE</code> messages will not be shown</li>
  <li><code class="highlighter-rouge">cache</code>: if assigned <code class="highlighter-rouge">TRUE</code> results will be cached to reuse in future rendering instances</li>
  <li><code class="highlighter-rouge">fig.height</code>: allows to specify height of figures in inches</li>
  <li><code class="highlighter-rouge">fig.width</code>: allows to specify width of figures in inches</li>
</ul>

<p>For more details on code chunk options see <a href="https://www.rstudio.com/wp-content/uploads/2015/03/rmarkdown-reference.pdf">here</a>.</p>

<h3 id="learning-markdown">Learning Markdown</h3>

<p>The basic syntax of Markdown and derivatives like kramdown is extremely easy to learn. Rather
than providing another introduction on this topic, here are some useful sites for learning Markdown:</p>

<ul>
  <li><a href="https://guides.github.com/features/mastering-markdown/">Markdown Intro on GitHub</a></li>
  <li><a href="https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet">Markdown Cheet Sheet</a></li>
  <li><a href="http://rmarkdown.rstudio.com/authoring_basics.html">Markdown Basics from RStudio</a></li>
  <li><a href="http://www.rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf">R Markdown Cheat Sheet</a></li>
  <li><a href="http://kramdown.gettalong.org/syntax.html">kramdown Syntax</a></li>
</ul>

<h3 id="tables">Tables</h3>

<p>There are several ways to render tables. First, they can be printed within the R code chunks. Second, 
much nicer formatted tables can be generated with the functions <code class="highlighter-rouge">kable</code>, <code class="highlighter-rouge">pander</code> or <code class="highlighter-rouge">xtable</code>. The following
example uses <code class="highlighter-rouge">kable</code> from the <code class="highlighter-rouge">knitr</code> package.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">knitr</span><span class="p">)</span><span class="w">
</span><span class="n">kable</span><span class="p">(</span><span class="n">iris</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">12</span><span class="p">,])</span><span class="w">
</span></code></pre>
</div>

<p>Sepal.Length   Sepal.Width   Petal.Length   Petal.Width  Species 
————-  ————  ————-  ————  ——–
          5.1           3.5            1.4           0.2  setosa<br />
          4.9           3.0            1.4           0.2  setosa<br />
          4.7           3.2            1.3           0.2  setosa<br />
          4.6           3.1            1.5           0.2  setosa<br />
          5.0           3.6            1.4           0.2  setosa<br />
          5.4           3.9            1.7           0.4  setosa<br />
          4.6           3.4            1.4           0.3  setosa<br />
          5.0           3.4            1.5           0.2  setosa<br />
          4.4           2.9            1.4           0.2  setosa<br />
          4.9           3.1            1.5           0.1  setosa<br />
          5.4           3.7            1.5           0.2  setosa<br />
          4.8           3.4            1.6           0.2  setosa</p>

<h3 id="figures">Figures</h3>

<p>Plots generated by the R code chunks in an R Markdown document can be automatically 
inserted in the output file. The size of the figure can be controlled with the <code class="highlighter-rouge">fig.height</code>
and <code class="highlighter-rouge">fig.width</code> arguments.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span><span class="w">
</span><span class="n">dsmall</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">diamonds</span><span class="p">[</span><span class="n">sample</span><span class="p">(</span><span class="n">nrow</span><span class="p">(</span><span class="n">diamonds</span><span class="p">),</span><span class="w"> </span><span class="m">1000</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">
</span><span class="n">ggplot</span><span class="p">(</span><span class="n">dsmall</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">color</span><span class="p">,</span><span class="w"> </span><span class="n">price</span><span class="o">/</span><span class="n">carat</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">geom_jitter</span><span class="p">(</span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">I</span><span class="p">(</span><span class="m">1</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<p><img src="Rbasics_files/figure-html/some_jitter_plot-1.png" alt="" /><!-- --></p>

<p>Sometimes it can be useful to explicitly write an image to a file and then insert that 
image into the final document by referencing its file name in the R Markdown source. For 
instance, this can be useful for time consuming analyses. The following code will generate a 
file named <code class="highlighter-rouge">myplot.png</code>. To insert the file  in the final document, one can use standard 
Markdown or HTML syntax, <em>e.g.</em>: <code class="highlighter-rouge">&lt;img src="myplot.png"/&gt;</code>.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">png</span><span class="p">(</span><span class="s2">"myplot.png"</span><span class="p">)</span><span class="w">
</span><span class="n">ggplot</span><span class="p">(</span><span class="n">dsmall</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">color</span><span class="p">,</span><span class="w"> </span><span class="n">price</span><span class="o">/</span><span class="n">carat</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">geom_jitter</span><span class="p">(</span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">I</span><span class="p">(</span><span class="m">1</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">))</span><span class="w">
</span><span class="n">dev.off</span><span class="p">()</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## png 
##   2
</code></pre>
</div>
<center><img title="some_title" src="myplot.png" /></center>

<h3 id="inline-r-code">Inline R code</h3>

<p>To evaluate R code inline, one can enclose an R expression with a single back-tick
followed by <code class="highlighter-rouge">r</code> and then the actual expression.  For instance, the back-ticked version 
of ‘r 1 + 1’ evaluates to 2 and ‘r pi’ evaluates to 3.1415927.</p>

<h3 id="mathematical-equations">Mathematical equations</h3>

<p>To render mathematical equations, one can use standard Latex syntax. When expressions are 
enclosed with single <code class="highlighter-rouge">$</code> signs then they will be shown inline, while 
enclosing them with double <code class="highlighter-rouge">$$</code> signs will show them in display mode. For instance, the following 
Latex syntax <code class="highlighter-rouge">d(X,Y) = \sqrt[]{ \sum_{i=1}^{n}{(x_{i}-y_{i})^2} }</code> renders in display mode as follows:</p>

<script type="math/tex; mode=display">d(X,Y) = \sqrt[]{ \sum_{i=1}^{n}{(x_{i}-y_{i})^2} }</script>

<h3 id="citations-and-bibliographies">Citations and bibliographies</h3>

<p>Citations and bibliographies can be autogenerated in R Markdown in a similar
way as in Latex/Bibtex. Reference collections should be stored in a separate
file in Bibtex or other supported formats. To cite a publication in an R Markdown 
script, one uses the syntax <code class="highlighter-rouge">[@&lt;id1&gt;]</code> where <code class="highlighter-rouge">&lt;id1&gt;</code> needs to be replaced with a 
reference identifier present in the Bibtex database listed in the metadata section 
of the R Markdown script  (<em>e.g.</em> <code class="highlighter-rouge">bibtex.bib</code>). For instance, to cite Lawrence et al. 
(2013), one  uses its reference identifier (<em>e.g.</em> <code class="highlighter-rouge">Lawrence2013-kt</code>) as <code class="highlighter-rouge">&lt;id1&gt;</code> [@Lawrence2013-kt]. 
This will place the citation inline in the text and add the corresponding
reference to a reference list at the end of the output document. For the latter a 
special section called <code class="highlighter-rouge">References</code> needs to be specified at the end of the R Markdown script.
To fine control the formatting of citations and reference lists, users want to consult this 
the corresponding <a href="http://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html">R Markdown page</a>.
Also, for general reference management and outputting references in Bibtex format <a href="https://paperpile.com/features">Paperpile</a> 
can be very helpful.</p>

<h1 id="session-info">Session Info</h1>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">sessionInfo</span><span class="p">()</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## R version 3.2.3 (2015-12-10)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 14.04.3 LTS
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
##  [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
## [10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  utils     datasets  grDevices methods   base     
## 
## other attached packages:
## [1] ggplot2_2.0.0   limma_3.26.3    BiocStyle_1.8.0
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.3      codetools_0.2-14 digest_0.6.9     plyr_1.8.3       grid_3.2.3      
##  [6] gtable_0.1.2     formatR_1.2.1    magrittr_1.5     evaluate_0.8     scales_0.3.0    
## [11] stringi_1.0-1    rmarkdown_0.9.2  tools_3.2.3      stringr_1.0.0    munsell_0.4.2   
## [16] yaml_2.1.13      colorspace_1.2-6 htmltools_0.3    knitr_1.12
</code></pre>
</div>

<h1 id="references">References</h1>



<div class="tags">
    
</div>

    

</div>

<hr class="shaded"/>

<footer>
            <div class="row">
                <div class="col-lg-12 footer">
               &copy;2016 Your company. All rights reserved. <br />
 Site last generated: Jul 10, 2016 <br />
<p><img src="images/company_logo.png" alt="Company logo"/></p>
                </div>
            </div>
</footer>


    </div>
    <!-- /.row -->
</div>
<!-- /.container -->
    </div>
    <script>
    jQuery(function ($) {
        $('.post-content').annotator()
                .annotator('setupPlugins', {tokenUrl: 'http://example.com/api/token'})
    });
    </script>
</body>

</html>